src/hb-ot-layout.cc

```diff
@@ -833,13 +849,12 @@ hb_ot_layout_collect_features (hb_face_t      *face,
       if (hb_ot_layout_table_find_script (face,
 					  table_tag,
 					  *scripts,
-					  &script_index))
-        _hb_ot_layout_collect_features_languages (face,
-                                                  table_tag,
-                                                  script_index,
+					  &script_index)
+          && context.visit (&context.gsubgpos.get_script (script_index)))
```

Comment:
I don't like that we are once calling get_script to check, and then call a function with script_index.  Please use the function chain to take the script object and next the langsys object, instead of indices, so we look them up once.  And I would move the "visit" check into the next function instead of before calling it.

Also, "visit" name is misleading since it suggests it does the actual visitation.  "may_visit" or something is better.  Or invert it as "visited"...